src/state/melmint.rs

```diff
@@ -147,88 +229,90 @@ fn process_swaps<C: ContentAddrStore>(mut state: State<C>) -> State<C> {
-                    Ratio::new(swap.outputs[0].value.0, total_lefts),
-                ))
-                    .min(MAX_COINVAL);
-            } else {
-                swap.outputs[0].denom = pool.left;
-                swap.outputs[0].value = CoinValue(multiply_frac(
-                    left_withdrawn,
-                    Ratio::new(swap.outputs[0].value.0, total_rights),
-                ))
-                    .min(MAX_COINVAL);
-            }
-            state.coins.insert_coin(
-                correct_coinid,
-                CoinDataHeight {
-                    coin_data: swap.outputs[0].clone(),
-                    height: state.height,
-                },
-                state.tip_906(),
-            );
-        });
-
-        state.pools.insert(*pool, pool_state);
+        let mut relevant_swaps: Vec<Transaction> = transactions_for_pool(&swap_reqs, pool);
+        process_swaps_for_single_pool(pool, &mut state, &mut relevant_swaps);
     });
 
     state
 }
 
+fn process_deposits_for_single_pool<C: ContentAddrStore>(
+    pool: &PoolKey,
+    state: &mut State<C>,
+    deposits: &mut Vec<Transaction>,
+) {
+    // sum up total lefts and rights
+    let total_lefts: u128 = deposits
+        .iter()
+        .map(|tx| tx.outputs[0].value.0)
+        .fold(0u128, |a, b| a.saturating_add(b));
+
+    let total_rights: u128 = deposits
+        .iter()
+        .map(|tx| tx.outputs[1].value.0)
+        .fold(0u128, |a, b| a.saturating_add(b));
+
+    let total_mtsqrt = total_lefts.sqrt().saturating_mul(total_rights.sqrt());
+    // main logic here
+    let total_liqs = if let Some(mut pool_state) = state.pools.get(pool).0 {
+        let liq = pool_state.deposit(total_lefts, total_rights);
+        state.pools.insert(*pool, pool_state);
+        liq
+    } else {
+        let mut pool_state = PoolState::new_empty();
+        let liq = pool_state.deposit(total_lefts, total_rights);
+        state.pools.insert(*pool, pool_state);
+        liq
+    };
+    // divvy up the liqs
+    deposits.iter_mut().for_each(|deposit| {
+        let original_tx = deposit.clone();
+        let my_mtsqrt = deposit.outputs[0]
+            .value
+            .0
+            .sqrt()
+            .saturating_mul(deposit.outputs[1].value.0.sqrt());
+        deposit.outputs[0].denom = pool.liq_token_denom();
+        deposit.outputs[0].value =
+            multiply_frac(total_liqs, Ratio::new(my_mtsqrt, total_mtsqrt)).into();
+        log::debug!(
+            "added {} total liquidity out of {}!",
+            deposit.outputs[0].value,
+            total_liqs
+        );
+        state.coins.insert_coin(
+            original_tx.output_coinid(0),
+            CoinDataHeight {
+                coin_data: deposit.outputs[0].clone(),
+                height: state.height,
+            },
+            state.tip_906(),
+        );
+        if (state.network == NetID::Mainnet || state.network == NetID::Testnet)
+            && state.height.0 < 978392
```

Comment:
@nullchinchilla I'd like to add some sort of note/documentation for this -- just leaving a note here for now